### PR TITLE
Attributes

### PR DESCRIPTION
**Pull request by Canop**
Replace the Vec<Attribute> with a bitset as proposed in #376

### CODE REVIEW

File: src/style/attributes.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    }
+}
+
+impl Attributes {
+    #[inline(always)]
+    pub fn push(&mut self, attribute: Attribute) {
+        self.0 |= attribute.bytes();
+    }
+    #[inline(always)]
+    pub fn remove(&mut self, attribute: Attribute) {
```
Comment (TimonPost) [inline src/style/attributes.rs]: What about `unset` in the sense of undoing the `set` operation, also to be consistent with `set` name.

File: src/style/attributes.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+    #[inline(always)]
+    pub fn has(&self, attribute: Attribute) -> bool {
+        self.0 & attribute.bytes() != 0
+    }
+    #[inline(always)]
+    pub fn extend(&mut self, attributes: &Attributes) {
+        self.0 = self.0 | attributes.0;
+    }
+    /// tells whether there's absolutely no attribute
```
Comment (TimonPost) [inline src/style/attributes.rs]:
```suggestion
    /// Returns whether there is no attribute set.
```

Comment (TimonPost) [reply src/style/attributes.rs]: In crossterm, I have plural verb + operation description.

File: src/style/ansi.rs [hunk 3]
```diff
@@ -21,7 +21,17 @@ pub(crate) fn set_bg_csi_sequence(bg_color: Color) -> String {
...
 pub(crate) fn set_attr_csi_sequence(attribute: Attribute) -> String {
-    format!(csi!("{}m"), attribute as i16)
+    format!(csi!("{}m"), attribute as u64)
+}
+
+pub(crate) fn set_attrs_csi_sequence(attributes: Attributes) -> String {
+    let mut s = String::new();
+    for attr in Attribute::iterator() {
+        if attributes.has(attr) {
+            s.push_str(&format!("\x1B[{}m", attr as u64));
```
[the remark this answers is not in the document]
Comment (Canop) [reply src/style/ansi.rs]: Unless I'm mistaken, this means another call and another string allocation. I don't think we should do it.

File: src/style/attributes.rs (hunk 1 above)
Comment (Canop) [reply src/style/attributes.rs]: I'm OK with that

File: src/style/content_style.rs [hunk 4]
```diff
@@ -51,7 +51,7 @@ impl ContentStyle {
     /// You can add more attributes by calling this method multiple times.
     #[inline]
     pub fn attribute(mut self, attr: Attribute) -> ContentStyle {
```
Comment (TimonPost) [inline src/style/content_style.rs]: Would it be an idea, to create a method that accepts `Attributes` as well?

File: src/style/attributes.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
+use crate::style::Attribute;
+use std::ops::{BitAnd, BitOr, BitXor};
+
+/// a bitset for all possible attributes
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct Attributes(u64);
```
Comment (TimonPost) [inline src/style/attributes.rs]: Do we really need a u64 for this? We can go way smaller right? Even something like a u8 would already be enough. Or am I mistaken?

File: src/style/ansi.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (TimonPost) [reply src/style/ansi.rs]: It uses `concat!(`. Which means it will be concatenated at runtime right?

[the remark this answers is not in the document]
Comment (Canop) [reply src/style/ansi.rs]: You might be right (assuming you mean at compile-time).

File: src/style/attributes.rs (hunk 5 above)
Comment (Canop) [reply src/style/attributes.rs]: There are 24 values. So a u8 can't store them all.
A u32 would be enough but then we need another implementation for `Attribute.bytes()`.
A possibility would be to remove the SGR indexes from the enum and hardcode them in the unix implementation of SetAttribute and SetAttributes. Which would probably not be dirtier.

File: src/style/content_style.rs (hunk 4 above)
Comment (Canop) [reply src/style/content_style.rs]: probably. Calling extend.

File: src/style/ansi.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (TimonPost) [reply src/style/ansi.rs]: Yes that is what I meant

File: src/style/attributes.rs [hunk 6]
```diff
@@ -61,26 +61,34 @@ impl BitXor for Attributes {
 }
 
 impl Attributes {
+    /// Sets the attribute.
+    /// If it's already set this does nothing.
```
Comment (TimonPost) [inline src/style/attributes.rs]:
```suggestion
    /// If it's already set, this does nothing.
```

File: src/style/content_style.rs (hunk 4 above)
Comment (TimonPost) [reply src/style/content_style.rs]: Oke

**Comment by TimonPost** [comment]
There are some clippy errrors. If those are resolved, this PR can be merged. Thanks again!

**Comment by TimonPost** [comment]
Awesome!

**Pull request merged**